File: sol/solution/0101-0200/0106/Solution.kt
/**
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    private var postIndex: Int = 0
    private val inMap = mutableMapOf<Int, Int>()    
    fun buildTree(inorder: IntArray, postorder: IntArray): TreeNode? {
        inorder.forEachIndexed { i, v -> inMap[v] = i }
        postIndex = postorder.size - 1
        return buildTreeHelper(inorder, postorder, 0, inorder.size - 1)
    }
    private fun buildTreeHelper(inorder: IntArray, postorder: IntArray, inStart: Int, inEnd: Int): TreeNode? {
        if (inStart > inEnd) return null
        val root = TreeNode(postorder[postIndex--])
        val inIndex = inMap[root.`val`]!!
        root.right = buildTreeHelper(inorder, postorder, inIndex + 1, inEnd)
        root.left = buildTreeHelper(inorder, postorder, inStart, inIndex - 1)
        return root
    }
}

File: sol/solution/0101-0200/0106/Solution.dart
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *   int val;
 *   TreeNode? left;
 *   TreeNode? right;
 *   TreeNode([this.val = 0, this.left, this.right]);
 * }
 */
class Solution {
  int postIndex = 0;
  Map<int, int> inMap = {};
  TreeNode? buildTree(List<int> inorder, List<int> postorder) {
    for (int i = 0; i < inorder.length; i++) {
      inMap[inorder[i]] = i;
    }
    postIndex = postorder.length - 1;
    return buildTreeHelper(inorder, postorder, 0, inorder.length - 1);
  }
  TreeNode? buildTreeHelper(List<int> inorder, List<int> postorder, int inStart, int inEnd) {
    if (inStart > inEnd) return null;
    TreeNode root = TreeNode(postorder[postIndex--]);
    int inIndex = inMap[root.val]!;
    root.right = buildTreeHelper(inorder, postorder, inIndex + 1, inEnd);
    root.left = buildTreeHelper(inorder, postorder, inStart, inIndex - 1);
    return root;
  }
}